### ISSUE TITLE
[BUG] bwl: nl80211_client::get_radio_info missing channels on RDK-B

### ISSUE THREAD

**Issue by itayx** [prplfoundation/prplMesh#1154]
**Describe the bug** -
as described in #1046, channels with the flag NO-IR still get filtered in the get_radio_info function.

**Comment by arnout** [comment]
@itayx Could you include the output of `read_supported_channels` and `iw phy0 info` like you did in #1046?

It looks like #1053 didn't completely fix #1046. Now the channels are mostly correct, but DFS channels are still missing.

The reason is that RDK-B uses a different version of the driver, and there the radar channels are masked as NO_IR. This means it is not allowed to configure an AP on that channel, but it is allowed to configure a STA. Since the supported_channels_list corresponds to the channels that can be used as an AP, the current behaviour appears to be correct at first sight.

However, it turns out that hostapd has an exception for this situation: if a channel is marked as a radar channel (`NL80211_FREQUENCY_ATTR_RADAR`) and DFS is enabled in hostapd, then the NO_IR bit is not considered. So in prplMesh we need to emulate the same behaviour.

After one-on-one discussion with Itay, we came to the following solution:
- nl80211_client_impl should keep track of the NO_IR attribute explicitly instead of simply ignoring the channel;
- the bwl implementation should take into account the `no_ir` and `is_dfs` fields of the channel list when constructing the supported_channel_list;
- for now, we assume that hostapd and driver *do* support DFS. If we ever run into a situation where this isn't the case, we'll handle it at that time. It's easy to notice it because hostapd will give an error when we try to select a DFS channel.

**Comment by kantera800** [comment]
>  DFS is enabled in hostapd

this flag **only** can be read from  hosatapd (patched msg) GET_HW_FEATURES that we are replacing ...

> * for now, we assume that hostapd and driver _do_ support DFS. If we ever run into a situation where this isn't the case, we'll handle it at that time. It's easy to notice it because hostapd will give an error when we try to select a DFS channel.

since we should do this assumption on every platform that we support i suggest: 
-> add method to bwl   `is_radar_detection_supported()` that returns true.
-> implement this method in DWPAL bwl  by reading from GET_HW_FEATURES 

@arnout @vitalybu  what do you think ?

**Comment by vitalybu** [comment]
> since we should do this assumption on every platform that we support i suggest:
> -> add method to bwl `is_radar_detection_supported()` that returns true.
> -> implement this method in DWPAL bwl by reading from GET_HW_FEATURES

What about `NL80211_FREQUENCY_ATTR_DFS_STATE`?
Why not rely on the information provided by the wlan driver?
That's what the hostapd does anyway: http://w1.fi/cgit/hostap/tree/src/drivers/driver_nl80211_capa.c#n1446

**Comment by kantera800** [comment]
> > since we should do this assumption on every platform that we support i suggest:
> > -> add method to bwl `is_radar_detection_supported()` that returns true.
> > -> implement this method in DWPAL bwl by reading from GET_HW_FEATURES
> 
> What about `NL80211_FREQUENCY_ATTR_DFS_STATE`?
> Why not rely on the information provided by the wlan driver?
> That's what the hostapd does anyway: http://w1.fi/cgit/hostap/tree/src/drivers/driver_nl80211_capa.c#n1446


great, even better idea. 
just verified that we get proper states in RDKB with `iw phy phy0 channels`

> 
> Band 2:
>         * 5280 MHz [56] 
>           Maximum TX power: 24.0 dBm
>           No IR
>           Radar detection
>           Channel widths: 20MHz HT40- HT40+ VHT80 VHT160
>           **DFS state: available** (for 24903 sec)
>           DFS CAC time: 60000 ms
>         * 5520 MHz [104] 
>           Maximum TX power: 24.0 dBm
>           No IR
>           Radar detection
>           Channel widths: 20MHz HT40- HT40+ VHT80 VHT160
>           **DFS state: usable** (for 24978 sec)
>           DFS CAC time: 60000 ms

both "available" and "usable" states are good for us .

**Comment by itayx** [comment]
well, currently we partly store the value of `NL80211_FREQUENCY_ATTR_DFS_STATE` in the `channel_info::radar_affected` which is a boolean value that equals true if the `NL80211_FREQUENCY_ATTR_DFS_STATE` is set to `NL80211_DFS_UNAVAILABLE`

**Comment by arnout** [comment]
DFS_STATE is the "dynamic" state, it doesn't say if DFS is supported or not. Channels that are marked as unavailable for DFS should *still* be in the supported channels list, but should get a preference value of 0 in the channel preference list.

**Comment by itayx** [comment]
> DFS_STATE is the "dynamic" state, it doesn't say if DFS is supported or not. Channels that are marked as unavailable for DFS should _still_ be in the supported channels list, but should get a preference value of 0 in the channel preference list.

So, the supported channels vector that is created using the `nl80211_client::get_radio_info`
Should:
- Have the NO_IR flag
- Store the value of `NL80211_FREQUENCY_ATTR_DFS_STATE`

**Comment by kantera800** [comment]
> DFS_STATE is the "dynamic" state, it doesn't say if DFS is supported or not. Channels that are marked as unavailable for DFS should _still_ be in the supported channels list, but should get a preference value of 0 in the channel preference list.

currently if its on the supported channel list it can be scanned.
@arnout are you sure we can scan "unavailable" channels?
also what about this state HOSTAPD_CHAN_DFS_UNKNOWN?

edit:
you are correct, dfs_status is not an indication for dfs support.

checked INTEL's hostapd code again ,
they use propriety flags for that .

```
    int dfs_enabled = hapd->iconf->**ieee80211h** &&
      (iface->drv_flags & WPA_DRIVER_FLAGS_RADAR);
```

# Enable IEEE 802.11h. This enables radar detection and DFS support if
# available. DFS support is required on outdoor 5 GHz channels in most countries
# of the world. This can be used only with ieee80211d=1.
# (default: 0 = disabled)
ieee80211h=1

/** Driver supports radar detection */
#define WPA_DRIVER_FLAGS_RADAR	

so, we still can go with my proposal
https://github.com/prplfoundation/prplMesh/issues/1154#issuecomment-616202989

or do you still prefer to add all non HOSTAPD_CHAN_DISABLED channels to the support list?

> Store the value of NL80211_FREQUENCY_ATTR_DFS_STATE

yes
*actually if its dynamic it makes no sense saving it (at least not as part of the supported channels)
saving `no_ir `and `is_dfs `is enough.

**Comment by arnout** [comment]
Let me summarize what I think we are converging to...
1.  nl80211 gives us 4 flags for each channel:
  - radar: this is a channel that overlaps with weather (and other) radar signals.
  - no_ir: this is a channel that we are not allowed to initiate transmission on.
  - no_ibss: this is a channel that we are not allowed to configure an AP on.
  - radar_detected: this is a radar channel on which a radar was detected and we need to perform CAC before we can use it.
2. We are actually interested only in 3 things:
  - Is the channel usable at all for an AP?
  - (if yes): do we need to do DFS to be able to use it?
  - (if yes): is the channel *currently* usable?

Given that the nl80211 stuff is nl80211-specific, I would extract *all* information we get from it and record it in the nl80211 structures. From this, the bwl should then extract the information that is actually used by prplMesh. So that means:
- is_dfs, no_ir and no_ibss booleans in the nl80211 structure;
- when converting to the supported_channels list, use the following logic:
  - if is_dfs, add it to the list with is_dfs set;
  - *else*, if no_ir or no_ibss, don't add it to the list.

**Comment by kantera800** [comment]
Good summary. 

You missed the "radar detection enabled" issue but i guess it can be addressed later since currently all our platforms support it.

**Comment by itayx** [comment]
Closing this issue as it was solved in #1156
